File: alert_app/android/app/src/main/java/io/flutter/plugins/alert_app/NotificationListenerService.java
package io.flutter.plugins.alert_app;

import android.app.Notification;
import android.content.Intent;
import android.os.Bundle;
import android.service.notification.StatusBarNotification;
import android.util.Log;
import java.util.HashMap;
import java.util.Map;
import com.example.alert_new.MainActivity;

public class NotificationListenerService extends android.service.notification.NotificationListenerService {
    private static final String TAG = "NotificationListener";
    
    private static final String[] UPI_APPS = {
        "com.google.android.apps.nbu.paisa.user", // Google Pay
        "com.phonepe.app",                        // PhonePe
        "net.one97.paytm",                       // Paytm
        "in.org.npci.upiapp",                    // BHIM
        "com.amazon.mShop.android.shopping",     // Amazon Pay
        "com.mobikwik_new",                      // MobiKwik
        "com.freecharge.android"                 // FreeCharge
    };

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG, "NotificationListenerService created");
        
        Intent serviceIntent = new Intent(this, NotificationForegroundService.class);
        startForegroundService(serviceIntent);
    }

    @Override
    public void onNotificationPosted(StatusBarNotification sbn) {
        try {
            String packageName = sbn.getPackageName();
            
            if (!isUpiApp(packageName)) {
                return;
            }

            Notification notification = sbn.getNotification();
            Bundle extras = notification.extras;
            
            if (extras != null) {
                String title = extras.getString(Notification.EXTRA_TITLE, "");
                String text = extras.getString(Notification.EXTRA_TEXT, "");
                String bigText = extras.getString(Notification.EXTRA_BIG_TEXT, "");
                
                String notificationText = bigText != null && !bigText.isEmpty() ? bigText : text;
                
                Log.d(TAG, "UPI Notification - Package: " + packageName + ", Title: " + title + ", Text: " + notificationText);
                
                sendNotificationToFlutter(packageName, title, notificationText);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error processing notification", e);
        }
    }

    @Override
    public void onNotificationRemoved(StatusBarNotification sbn) {
        // Handle notification removal if needed
    }

    private boolean isUpiApp(String packageName) {
        for (String upiApp : UPI_APPS) {
            if (upiApp.equals(packageName)) {
                return true;
            }
        }
        return false;
    }

    private void sendNotificationToFlutter(String packageName, String title, String text) {
        Map<String, String> data = new HashMap<>();
        data.put("packageName", packageName);
        data.put("title", title);
        data.put("text", text);
        data.put("timestamp", String.valueOf(System.currentTimeMillis()));
        
        // Send via static method in MainActivity
        try {
            MainActivity.sendNotificationEvent(data);
        } catch (Exception e) {
            Log.e(TAG, "Error sending notification to Flutter", e);
        }
    }
}

File: alert_app/android/app/src/main/kotlin/com/example/alert_new/NotificationListenerService.kt
package com.example.alert_new

import android.app.Notification
import android.content.Intent
import android.os.Bundle
import android.service.notification.NotificationListenerService
import android.service.notification.StatusBarNotification
import io.flutter.plugin.common.MethodChannel
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.dart.DartExecutor

class NotificationListenerService : NotificationListenerService() {
    
    private val upiPackages = setOf(
        "com.phonepe.app",
        "com.google.android.apps.nfc.payment", 
        "net.one97.paytm",
        "in.org.npci.upiapp",
        "in.amazon.mShop.android.shopping"
    )
    
    override fun onNotificationPosted(sbn: StatusBarNotification?) {
        super.onNotificationPosted(sbn)
        
        sbn?.let { notification ->
            if (upiPackages.contains(notification.packageName)) {
                processUpiNotification(notification)
            }
        }
    }
    
    private fun processUpiNotification(sbn: StatusBarNotification) {
        val extras = sbn.notification.extras
        val title = extras.getString(Notification.EXTRA_TITLE) ?: ""
        val text = extras.getString(Notification.EXTRA_TEXT) ?: ""
        val bigText = extras.getString(Notification.EXTRA_BIG_TEXT) ?: text
        
        // Send to Flutter
        val data = mapOf(
            "packageName" to sbn.packageName,
            "title" to title,
            "text" to bigText,
            "timestamp" to sbn.postTime
        )
        
        sendToFlutter(data)
    }
    
    private fun sendToFlutter(data: Map<String, Any>) {
        try {
            // Send via broadcast or method channel
            val intent = Intent("UPI_NOTIFICATION_RECEIVED")
            intent.putExtra("notification_data", data.toString())
            sendBroadcast(intent)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }
}